hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkRunClusteringCommitActionExecutor.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class SparkRunClusteringCommitActionExecutor<T extends HoodieRecordPayload<T>>
+    extends BaseSparkCommitActionExecutor<T> {
+
+  private static final Logger LOG = LogManager.getLogger(SparkRunClusteringCommitActionExecutor.class);
+  private final HoodieClusteringPlan clusteringPlan;
+
+  public SparkRunClusteringCommitActionExecutor(HoodieEngineContext context,
+                                                HoodieWriteConfig config, HoodieTable table,
+                                                String instantTime) {
+    super(context, config, table, instantTime, WriteOperationType.CLUSTER);
+    this.clusteringPlan = ClusteringUtils.getClusteringPlan(table.getMetaClient(), HoodieTimeline.getReplaceCommitRequestedInstant(instantTime))
+      .map(Pair::getRight).orElseThrow(() -> new HoodieClusteringException("Unable to read clustering plan for instant: " + instantTime));
+  }
+
+  @Override
+  public HoodieWriteMetadata<JavaRDD<WriteStatus>> execute() {
+    HoodieInstant instant = HoodieTimeline.getReplaceCommitRequestedInstant(instantTime);
+    // Mark instant as clustering inflight
+    table.getActiveTimeline().transitionReplaceRequestedToInflight(instant, Option.empty());
+    table.getMetaClient().reloadActiveTimeline();
+
+    JavaSparkContext engineContext = HoodieSparkEngineContext.getSparkContext(context);
+    // read rdd from input groups in plan
+    JavaRDD<WriteStatus> writeStatuses = clusteringPlan.getInputGroups().stream()
+        .map(inputGroup -> runClusteringForGroup(inputGroup, clusteringPlan.getStrategy().getStrategyParams()))
+        .reduce((rdd1, rdd2) -> engineContext.union(rdd1, rdd2)).orElse(engineContext.emptyRDD());
+    if (writeStatuses.isEmpty()) {
+      throw new HoodieClusteringException("Clustering plan produced 0 WriteStatus for " + instantTime + " #groups: " + clusteringPlan.getInputGroups().size());
+    }
+    // merge all write status
+    HoodieWriteMetadata<JavaRDD<WriteStatus>> writeMetadata = buildWriteMetadata(writeStatuses);
+    updateIndexAndCommitIfNeeded(writeStatuses, writeMetadata);
+    if (!writeMetadata.getCommitMetadata().isPresent()) {
+      HoodieCommitMetadata commitMetadata = CommitUtils.buildMetadata(writeStatuses.map(WriteStatus::getStat).collect(), writeMetadata.getPartitionToReplaceFileIds(),
+          extraMetadata, operationType, getSchemaToStoreInCommit(), getCommitActionType());
+      writeMetadata.setCommitMetadata(Option.of(commitMetadata));
+    }
+    return writeMetadata;
+  }
+
+  private JavaRDD<WriteStatus> runClusteringForGroup(HoodieClusteringGroup clusteringGroup, Map<String, String> strategyParams) {
```

Comment:
So if there are N clustering groups consisting of M file groups that are being converted into O file groups, we will have N different RDD's. There are some performance limitation of unioning multiple different RDDs from what I remember. Can you run a simple test a) take 10000 records, parallelize them with jsc to create 1 RDD and then write them out to a file b) take 10000 records, create 1 RDD for each record and then union them and write them out to a single file ?

Reply:
Is there an alternative to union? Should we collect individual RDD and merge lists? I did testing only with 2 groups so far. So, I didn't see big performance degradation. I'll look into testing large number of groups as separate task. (We may also don't want number of groups to be really high as we want clustering to be atomic operation. Doing large amount of data/groups can increase chance of failures and slow down entire process.)

Reply:
Yes, like we discussed offline, we should have a config to limit the number of clustering groups to avoid someone setting wrong configs for things like maxDataPerGroup and then result in a large number of RDDs to be unioned. I think we should still do a quick test to see if there is any degradation due to union before we land this diff so we are aware. It should be as simple as writing a test case that loops over code

TestUnionPerformance

long startTime = System.currentMillis();
RDD rdd = jsc.parallelize(Arrays.asList(record))
for (record in records)
{
rdd.union(jsc.parallelize(Arrays.asList(record)))
}
rdd.collect();
writeClient.bulkInsert(rdd);
sout("time taken with union" + System.currentMillis - starttime)

long startTime = System.currentMillis();
RDD rdd = jsc.parallelize(records)
rdd.collect();
writeClient.bulkInsert(rdd);
sout("time taken without union" + System.currentMillis - starttime)

Reply:
Ok, Changed to limit max groups to 50 (thats where we see significant degradation). I only applied for this strategy (because reduce is inside strategy code).